### ISSUE TITLE
Add support for variadic metavariables or ellipsis operator for post-processing

### ISSUE THREAD

**Issue by mschwager** [semgrep/semgrep#744]
**Is your feature request related to a problem? Please describe.**

I'd like to be able to obtain information about certain code patterns when the general form of the code is known, but the number is unknown.

For example, say I'd like to grab information about certain Python decorators:

```python
@dec1
@dec2(arg="foo")
@dec3
def func(arg):
    ...
```

Maybe I'd like to find when `@dec2` is used with `arg="foo"`.

First, consider that the decorators may be in any order. This means we have a combinatoric explosion for expressing all the different orders. Second, if we consider that there may be 10 different decorators, but only 3-4 are used in any given situation, then it becomes near-impossible to create a rule to describe all these scenarios. Think API routes with different authz and authn decorators to describe how they may be accessed.

Another common pattern where this is a problem is variadic information stored in a data structure such as a list or mapping. For example

```
routes = [
    {path: "/", authn: User, authz: Admin, handler: IndexHandler},
    {path: "/users", authn: User, authz: Admin, handler: UsersHandler},
    ...
]
```

Let's say I'd like to obtain the `path`, `authz`, and `handler` for each route, but the number of routes changes over time. The ellipsis operator is supported in arrays, but as far as I can tell it's not possible to pull out structured information from a variadic structure (e.g. the `path` from each object).

The common functionality here is pulling out structured information from a variadic number of code patterns.

**Describe the solution you'd like**

I'm not sure about the best way to express this. The general idea is as follows:

```
pattern: |
    ...@$DECORATORS...
    def $FUNC(...):
        ...
```

Then we can perform some post-processing on the information obtained from `$DECORATORS` and `$FUNC`. Similarly for objects:

```
routes = [
    { ..., path: $PATH, handler: $HANDLER, ... }
    ...
]
```

Or something like that. Maybe we need a new operator similar to the ellipsis operator to handle these cases. Another useful feature above would be saying `$DECORATORS` must contain decorator `@route`, and only then do we grab and return ALL other decorators for that function.

Basically the ask is a way to grab variadic code patterns that meet a certain structure. This is mostly a feature that would be used with additional post processing, for example considering how a route behaves when looking at all its decorators.

**Describe alternatives you've considered**

One alternative is trying to enumerate all possible code patterns that meet the general structure. E.g. enumerate all orderings of a certain group of decorators. This becomes tedious, if not impossible with larger amounts, and is still prone to missing certain cases (e.g. 10 choose 4 decorators). 

**Additional context**

None

**Comment by mschwager** [comment]
Another example from Java land, consider an API route like:

```java
    @POST
    @ROUTE("/users")
    public String users(@PathParam("name") String name,
                                @QueryParam("version") Integer version,
                                @QueryParam("id") String id,
                                @QueryParam("priority") @DefaultValue("0") Integer priority,
                                Map<String, Object> input) {
        return db.users(name, version, id, priority, input);
    }
```

Here the number of `@QueryParam`s and decorators will be variadic when looking at many different functions. So we'd like to grab the function name and a structured representation of all the query parameters and decorators for post-processing.

**Comment by DrewDennison** [comment]
similar to https://github.com/returntocorp/semgrep/issues/849

**Comment by sabrinabrogren** [comment]
Another example of where this may be useful is with writing patterns for Ruby routes.
```
Rails.application.routes.draw do
  resources :book, :image, :archive
end
```
It would be useful to be able to capture all of these resources.

**Comment by mschwager** [comment]
Still a thing

**Comment by clintgibler** [comment]
@mschwager already gave a great overview and description, but here's some more details about how I'm imagining it at least.

Here are some specific use case examples of what I'd like to be able to do, taking Java routes as an example:
https://semgrep.dev/s/clintgibler:java-routes

I'm being a bit handwavy about what the JSON results should look like and I'm not tied to it, it could just be raw AST, I'm just trying to convey what I'd want as a user.

## Extracting arguments/parameters

Sometimes it's useful to be able to extract all of the arguments to an annotation, parameters to a function definition, or arguments to a method call.

### Annotation Arguments

```java
@RequestMapping(method = RequestMethod.GET)
...
@RequestMapping(value = "/{name}", method = RequestMethod.POST)
```

Ideally I'd like to be able to have a pattern like like `@RequestMapping($...ARGS)` and `$...ARGS` will give me a representation like:

```json
{
  "method": "RequestMethod.GET"
}
```

for the first and for the second:

```json
{
  "method": "RequestMethod.POST",
  "value": "/{name}"
}
```

### Method Definitions / Function Calls

I imagine these would work similar to the above example, where the pattern `$FUNC($...ARGS)` called on

```java
public String func(String name, int age) {
   ...
}
```

would return

```json
[
  {
    "type": "String",
    "name": "name"
  },
  {
    "type": "int",
    "name": "age"
  },
]
```

Or @mschwager's example above:

```java
    public String users(@PathParam("name") String name,
                                @QueryParam("version") Integer version,
                                @QueryParam("id") String id,
                                @QueryParam("priority") @DefaultValue("0") Integer priority,
                                Map<String, Object> input)
```

```json
[
  {
    "annotation": { "name": "PathParam", "value": "name" },
    "type": "String",
    "name": "name"
  },
  {
    "annotation": { "name": "QueryParam", "value": "version" },
    "type": "Integer",
    "name": "version"
  },
  ...
]
```
Maybe add another field for default values?

## Extracting All Annotations

Sometimes it's useful to grab *all* of the annotations on a method, but you don't know beforehand what they are.

So running the pattern:
```python
$...ANNOTATIONS
def $FUNC(...):
  ...
```
  on

```python
@dec1
@dec2(arg="foo")
@dec3
def func(arg):
    ...
```

would yield:

```json
[
  {
    "name": "dec1"
  } ,
  {
    "name": "dec2",
    "arguments": [
      { "name": "arg", "value": "foo" }
    ]
  } ,
  {
    "name": "dec3"
  } ,
]
```

**Comment by na-ka-na** [comment]
+1 we want to use semgrep, but the first use case for us is - the right ordering of python decorators if multiple of them are used
